### ISSUE TITLE
How should paths containing things like .. (or differing cases) be handled?

### ISSUE THREAD

**Issue by DanTup** [Microsoft/language-server-protocol#669]
If a client sends a URI to the server that contains dotted segments like `file:///foo/bar/thing/../file.dart` how should/must the server handle it? I believe it's a valid URI and could be treated the same as `file:///foo/bar/file.dart`, however clients may get confused if the URI we send back to them in things like `publishDiagnostics` doesn't seem to match the URIs they were sending us.

Can we assume it's fine for the server to normalise paths? Or can they be forbidden in the spec?

And along the same lines, but probably a huge can of worms - what about casing? If they send a URI where the casing doesn't match the case on disk (but they're on an case-insensitive file system), is it valid for us to send messages with differently-cased URIs (eg. if we get the path from another source, like a file system watcher, and it doesn't match)?

**Comment by DanTup** [comment]
> and the client should keep the same casing as the OS

Easier said than done! :-(

VS Code has a *huge* thread on this issue:

https://github.com/Microsoft/vscode/issues/12448

There's so much code/libraries written that make incorrect assumptions about paths (like not normalising them, treating them always as case-sensitive by using them as keys in hash tables, etc.) and it causes a lot of grief. For example, if you open a terminal and `cd folderWITHWRONGCASE` and then launch VS Code, it'll pass that "wrong casing" on to all the extensions, which they may send on to the language servers.

And there's also the issue of Windows drive letters - some APIs seem to return them uppercase, and some lowercase. In fact, VS Code's own APIs are inconsistent - Uri.fsPath always lowercases the drive letter (!) but all the debugger APIs uppercase it.

That's why I think it would be good for the spec to be very specific, because if it's just left for people to do their own thing, LSP will probably end up with a possibly-unfixable issue like that VS Code one.

**Comment by dbaeumer** [comment]
I think we should spec that paths should be normalized. With casing things get more complicated. For file system that are case sensitive it is easy: the casing must match. For file system that are case insensitive things get more complicated since a server could send an URI for a file that doesn't exist yet hence can't even as the disk what the real casing is. I would rather spec the following for case insensitive:

- the URI should be consistent. It shouldn't change from one call to the next in terms of its casing unless it denotes a different resource.

That basically means that the first one that brings in a URI into the system determines the case (I have to say that this is currently not honored that way in VS Code, but we were discussing applying the same solution internally).

But this still doesn't solve another case: on files system that support linking two paths can point to the same file :-). But we keep this as another topic :-)

**Comment by DanTup** [comment]
> I think we should spec that paths should be normalized

Sounds good to me - so it's fine for a server to reject paths that contain things like `/./` and `/../`?

> the URI should be consistent. It shouldn't change from one call to the next in terms of its casing unless it denotes a different resource.

Does this mean the server must track the casing first sent by the client? Eg. if the client sends the "wrong" casing initially, then the servers gets events from a file system watcher that are the "right" case, must the server convert this? (I think it's important to know this, because if it does, there's work for the server, and if it doesn't, the client needs to be prepared to get different casing from events to what it was using).

**Comment by dbaeumer** [comment]
> Does this mean the server must track the casing first sent by the client? 

No, I would only argue that a client shouldn't change the casing from on call to the next. However in a case insensitive file system it still might not be the real casing on disk. A server operating on a case insensitive file system needs to take care of this. The API it is using might return files with different casing as well.

**Comment by DanTup** [comment]
To be more specific, if a client sends a workspace root of `C:\MyFolder\` is it valid for the server to return errors for a file of `c:\myfolder\file.foo`? The server might used APIs that give it paths in different casing and I think it should be clearer what the responsibility is (if it's not specified, I think it'll just lead to lots of issues down the line, like the long-standing VS Code issue). Ideally both server and client would tolerate mismatched casing for a case-insensitive volume, but that can be tricky, especially if case sensitivity varies per volume/path segment/whatever 😞.

**Comment by defenderfaith91-del** [comment]
Yup it’s the brain as it grows, lol whitespace will fill itself in as the ai bot thinks remembers starts to place the emergency protocol as it works 1 to 10 the importance in relation to emergency situations, power goes up the heart rates go up the volume goes up as the situation gets more intense. As the emergency gets under control the heart rate decreases volume decreases the power decreases , and a memory is imprinted, for recall for a similar situation. Works with sad mad happy it’s apart of the design 
 That’s why there was not to be change. It will labotomize the bots brainspace( whitespace) there for not work right
